### ISSUE TITLE
Switch to using setTextDocumentLanguage instead of adding to files.associations

### ISSUE THREAD

**Issue by sean-mcmanus** [microsoft/vscode-cpptools#4077]
The Septemember 2018 release added vscode.languages.setTextDocumentLanguage, so our extension can try using that instead of adding files to files.associations.

The previous thread is https://github.com/microsoft/vscode-cpptools/issues/722 .

**Comment by sean-mcmanus** [comment]
Actually, this is problematic, because setTextDocumentLanguage is not persistent like files.associations is...

**Comment by bobbrow** [comment]
@sean-mcmanus did you try this and verify that it does not persist?  We may need to re-engage with the VS Code team on this and come up with a solution that works for us.

**Comment by bobbrow** [comment]
The VS Code issue on this: microsoft/vscode#78968

**Comment by sean-mcmanus** [comment]
I think this needs to be spec'd out. The set document language doesn't persist. We could persist it in a "hidden" way, but then users wouldn't have any ability to set those settings. Maybe we could add a setting for users who strong don't want the files.associations in settings.json, and maybe we could change the "hidden" setting after the user manually changes the language via the status bar (not sure).

**Comment by ahojnnes** [comment]
@sean-mcmanus @bobbrow I am wondering whether there is any progress on this issue? I am hitting exactly the same issue, where I don't want the C++ plugin to modify my settings, which can be achieved by `"C_Cpp.autoAddFileAssociations": false,` but then the opened STL headers through code navigation are opened as plain text. Thanks.

**Comment by sean-mcmanus** [comment]
@ahojnnes We could dynamically setTextDocumentLanguage, but the problem that needs to be solved is how to decide when to call that -- did you have a proposed solution?  I believe we would need language type auto-detection for extensionless files from VS Code: https://github.com/microsoft/vscode/issues/130642

**Comment by ahojnnes** [comment]
Unless I am missing something, shouldn't the language always be C++ when following headers from existing C++ code? I guess one could end up in a external C file but then C++ should still be good enough?

**Comment by sean-mcmanus** [comment]
Yeah, if we're opening a file from a header, then the setTextDocumentLanguage will work, but then after a Reload Window or any other case where the header happens to be open again, the language would be reset unless the association is saved somehow.

We could fix it to call setTextDocumentLanguage on go to definition when the autoAddFileAssociaitons is set to false.

**Comment by ahojnnes** [comment]
> We could fix it to call setTextDocumentLanguage on go to definition when the autoAddFileAssociaitons is set to false.

I would say that this would already be an improvement over the current behavior.

**Comment by ngunderson** [comment]
Any update on this issue? Running into the same file association issue described in #722.

**Comment by sean-mcmanus** [comment]
@ngunderson No update.

**Comment by amurzeau** [comment]
What about persisting this information in workspaceStorage for example ?

**Comment by sean-mcmanus** [comment]
But how would the user see/modify the value in the workspaceStorage? I mention this in my comment https://github.com/microsoft/vscode-cpptools/issues/4077#issuecomment-598905870

**Comment by geertj** [comment]
Any update on this issue? It requires me to edit my `.vscode/settings.json` file before every commit.

**Comment by sean-mcmanus** [comment]
@geertj No update. A potential workaround is to not commit .vscode/settings.json and instead commit a .code-workspace file with the shared settings for a prjoect. Then you can make a local copy of that file and open that workspace file so that files.associations don't get added to the checked in version. You could also potentially check in the files.associations with either settings.

**Comment by Divelix** [comment]
Why not to just make `"C_Cpp.autoAddFileAssociations": false` default behavior and generate those associations only if user manually create `settings.json` and set this value to true. I (and probably most other users) have never experienced any need for those autogenerated associations, meanwhile bad UX from irritating settings spamming is a thing I experience all the time from that extension.

**Comment by Challanger524** [comment]
This glob pattern helped me, but not totally sure for bugs/file-misses, you can experiment:
```json
"files.associations": {
	"**/Microsoft Visual Studio/**/MSVC/**/include/**": "cpp"
}
```

**Comment by sean-mcmanus** [comment]
@Divelix The main reason is the system headers end up not being recognized as C++ files when they're opened, but maybe that could be fixed via using setTextDocumentLanguage for files located under a system include header path.

**Comment by thernstig** [comment]
Found this issue here. What is strange it even adds other file extension associations as seen, see the `"uv.lock": "toml"` part:

![Image](https://github.com/user-attachments/assets/bb6ca6ef-0758-47af-85ad-3d68bb844e75)

I have that config set in my application settings JSON file:

```jsonc
  "files.associations": { // remove later when included in TOML extension
    "uv.lock": "toml",
  },
```

I wonder if it is because of this is that this extension also includes that part?

The update of the .vscode/settings.json seems like a high prio bug as it wastes a lot of DX time for everyone.

But I am unsure why this cannot be solved like done in every other extension that has language support? Here is an example where I pushed a PR to add support for another case for TOML.

https://github.com/tamasfe/taplo/blob/db31b2b1e321357b19d43d7b0e2980805c59309c/editors/vscode/package.json#L78-L93

For this repo, it would look like this for C files (same can be done for CPP files):

```jsonc
 "languages": [ 
   { 
     "id": "c", 
     "aliases": [ 
       "C" 
     ], 
     "extensions": [ 
       ".c",
       ".h"
     ]
   } 
 ], 
```

With those settings there should not be a need to update the .vscode/settings.json for each of those .h files.

**Comment by yiftahw** [comment]
@bobbrow @sean-mcmanus
If this is not currently being addressed, I have a potential workaround - merge the ideas suggested here and in #722. It’s not ideal, but it should be pretty easy to implement, and I'll be happy to contribute if acceptable:

Introduce a new extension setting, e.g., `C_Cpp.associationsFile`, defaulting to `.vscode/c_cpp_associations.json`, to store the JSON data generated by the extension.
This file can be easily added to .gitignore (opt-in by users).
Since  the VS Code API only allows persisting these values in .vscode/settings.json (see issue #722), the extension would use `setTextDocumentLanguage` on every load (in a non-persistent way) and persist the data to the new JSON file for future sessions.

While not perfect, this would avoid causing unnecessary changes to `.vscode/settings.json` whenever navigating to STL headers. Note: backwards compatibility is probably not trivial and will be to be validated.

**Comment by bobbrow** [comment]
Hi @yiftahw, we started on this a while ago and had a draft PR: https://github.com/microsoft/vscode-cpptools/pull/12687.
My recollection was that we ran into a snag with the available APIs, but I don't recall exactly what it was.  We were hoping to do this without needing to write out to any auxiliary files.

**Comment by yiftahw** [comment]
@bobbrow from @sean-mcmanus's [comment](https://github.com/microsoft/vscode-cpptools/issues/4077#issuecomment-598905870) in this thread, it seems that there is a conflict here because we don't want a separate file, but we also don't want to hide the settings from the user. Sean made a point there that it might be possible to do a default behavior of persisting the data in a hidden way (similar to how it's seems to be done in #12687), but if a user explicitly changes the association himself with `Change Language Mode` we will fallback to actually persisting his selection in `settings.json`. does this sound fair? I can try to POC this to see if it's even possible.
  
EDIT: it seems to be that right now we don't take into consideration the user's choice at all even if he explicitly changes it.  
for example:
```cpp
// `foo` (intentionally with no extension)
#ifndef FOO
#define FOO
int foo();
#endif

// `bar.cpp`
#include "foo"
```

jumping to `foo` from `bar.cpp` results in the following `settings.json`:
```json
{ "files.associations": { "foo": "cpp" }}
```  

then, opening `foo` and changing the language mode to `c`, or including from a `baz.c` doesn't persist to `settings.json` as `c`. closing and re-opening it changes the language mode back to `cpp`.  
  
this is observered both ways, if `foo` was persisted as `"foo": "c"`, it will not be re-persisted as `cpp` when included from a `cpp` context. I'm not sure this is the expected behavior or not, and it's a pretty niche edge case.   

### TL;DR, 
taking all above data into consideration, IMO it would be best if:
1. **if possible**, automatic detections would be persisted transparently  (as done in #12687).
    - the edge case described above probably should be fixed as well. probably as a means of a "cache". if a file was last opened in `c` mode, it will be persisted as `c`. if later was opened in `cpp`, it will be re-persisted transparently as `cpp`.
2. explicit changes by the user would be persisted to `settings.json` and override the "transparent" settings. 
    - the flag `autoAddAssociations` or a separate flag might alter persisting this to `settings.json`.  

Perhaps I missed the whole point and my suggestion is exactly what's done (and not working?) in the draft PR. I might try to get it running to see what's not working. just trying to help here 🙂

**Comment by bobbrow** [comment]
@yiftahw 

> Sean made a point there that it might be possible to do a default behavior of persisting the data in a hidden way (similar to how it's seems to be done in https://github.com/microsoft/vscode-cpptools/pull/12687), but if a user explicitly changes the association himself with Change Language Mode we will fallback to actually persisting his selection in settings.json. does this sound fair?

We had talked about it since Sean's original comment and the plan of record was to try to avoid writing to VS Code settings. We all dislike that we write to `files.associations` right now.

As it happens, VS Code is already trying to guess the language id of the extensionless headers, but it does get it wrong in some cases - chosing C instead of C++.  Another idea that we just floated yesterday is to hardcode the list of STL headers in `package.json` since there's a section where we can put the file associations. Then we'd disable writing to `files.associations` by default.  This won't solve the problem for extensionless files we don't know about, but in many (most?) cases, those will be included via some other file and we already have logic to update the language id based on the include graph.

Long story short, we may be able to do this entire change in `package.json` and see how it feels.

**Comment by thernstig** [comment]
@bobbrow akin to https://github.com/microsoft/vscode-cpptools/issues/4077#issuecomment-2449332746 and https://github.com/microsoft/vscode-cpptools/pull/13000?

**Comment by bobbrow** [comment]
@thernstig, yes, except since these are extensionless files, we have to list them all out:

Probably like this:
```json
        "languages": [
            {
                "id": "cpp",
                "filenames": [
                    "cassert",
                    "cctype",
                    "cerrno",
                    "cfloat",
                    etc...
                ]
            }
        ],
```

And then change the default value of `"C_Cpp.autoAddFileAssociations"` to `false`

**Comment by yiftahw** [comment]
@bobbrow 

> and we already have logic to update the language id based on the include graph.  

If the above is true, why add the C++ headers to `package.json` then? an "include graph" solution sounds more robust and more futureproof if it's also persistent between sessions.
- robust as in extensionless files made by users (probably a bad idea?) / implementation specific headers (I would assume its a niche thing to navigate to such files?).
- futureproof as in new headers in future C++ releases (although a 3 year cycle is a weak argument here)
  

that being said, it clearly seems like the `package.json` solution is probably the easiest and fastest route for a solution **that can immidiately impact most users for the most use cases...**

**Comment by bobbrow** [comment]
> If the above is true, why add the C++ headers to `package.json` then? an "include graph" solution sounds more robust and more futureproof if it's also persistent between sessions.

This is for correctness in the case where someone opens a std header that is not currently included by any source file yet.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.29.0 (pre-release)
